File: Windows/programmers/HyperFlash.cs
using System;
using System.Linq;
using Microsoft.VisualBasic;

namespace FlashcatUSB {
    public class HF_Programmer : FlashcatUSB.MemoryDeviceUSB {
        private FlashcatUSB.USB.FCUSB_DEVICE FCUSB;

        public FlashcatUSB.FlashMemory.HYPERFLASH MyFlashDevice { get; set; }  // Contains the definition for the parallel mode device that is connected
        public FlashcatUSB.USB.DeviceStatus MyFlashStatus { get; set; } = FlashcatUSB.USB.DeviceStatus.NotDetected;

        public event PrintConsoleEventHandler PrintConsole;

        public delegate void PrintConsoleEventHandler(string msg);

        public event SetProgressEventHandler SetProgress;

        public delegate void SetProgressEventHandler(int percent); // Not used

        public HF_Programmer(FlashcatUSB.USB.FCUSB_DEVICE parent_if) {
            FCUSB = parent_if;
        }

        public bool DeviceInit() {
            FCUSB.USB_VCC_OFF();
            MyFlashDevice = null;
            this.EXPIO_SETUP_USB(FlashcatUSB.MEM_PROTOCOL.HYPERFLASH);
            FCUSB.USB_VCC_ON(FlashcatUSB.MainApp.MySettings.VOLT_SELECT);
            FlashcatUSB.Utilities.Main.Sleep(200);
            var HF_DETECT = DetectFlash();
            if (HF_DETECT.Successful) {
                string chip_id_str = Conversion.Hex(HF_DETECT.MFG).PadLeft(2, '0') + Conversion.Hex(HF_DETECT.ID1).PadLeft(8, '0');
                PrintConsole?.Invoke(string.Format(FlashcatUSB.MainApp.RM.GetString("ext_connected_chipid"), chip_id_str));
                var device_matches = FlashcatUSB.MainApp.FlashDatabase.FindDevices(HF_DETECT.MFG, HF_DETECT.ID1, 0, FlashcatUSB.FlashMemory.MemoryType.HYPERFLASH);
                if (device_matches is object && device_matches.Count() > 0) {
                    if (MyFlashDevice is null)
                        MyFlashDevice = (FlashcatUSB.FlashMemory.HYPERFLASH)device_matches[0];
                    PrintConsole?.Invoke(string.Format(FlashcatUSB.MainApp.RM.GetString("flash_detected"), MyFlashDevice.NAME, Strings.Format((object)MyFlashDevice.FLASH_SIZE, "#,###")));
                    MyFlashStatus = FlashcatUSB.USB.DeviceStatus.Supported;
                } else {
                    PrintConsole?.Invoke(FlashcatUSB.MainApp.RM.GetString("unknown_device_email"));
                    MyFlashStatus = FlashcatUSB.USB.DeviceStatus.NotSupported;
                }

                return true;
            } else {
                FlashcatUSB.MainApp.GUI.PrintConsole(FlashcatUSB.MainApp.RM.GetString("ext_not_detected"));
                MyFlashStatus = FlashcatUSB.USB.DeviceStatus.NotDetected;
                return false;
            }
        }

        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        internal string DeviceName {
            get {
                switch (MyFlashStatus) {
                    case FlashcatUSB.USB.DeviceStatus.Supported: {
                            return MyFlashDevice.NAME;
                        }

                    case FlashcatUSB.USB.DeviceStatus.NotSupported: {
                            return Conversion.Hex(MyFlashDevice.MFG_CODE).PadLeft(2, '0') + " " + Conversion.Hex(MyFlashDevice.ID1).PadLeft(4, '0');
                        }

                    default: {
                            return FlashcatUSB.MainApp.RM.GetString("no_flash_detected");
                        }
                }
            }
        }

        internal long DeviceSize {
            get {
                return MyFlashDevice.FLASH_SIZE;
            }
        }

        public byte[] ReadData(long logical_address, long data_count, FlashcatUSB.FlashMemory.FlashArea memory_area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            return ReadBulk_NOR((uint)logical_address, (uint)data_count);
        }

        public bool SectorErase(uint sector_index, FlashcatUSB.FlashMemory.FlashArea memory_area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            try {
                if (sector_index == 0L && SectorSize(0U) == MyFlashDevice.FLASH_SIZE) {
                    return EraseDevice(); // Single sector, must do a full chip erase instead
                } else {
                    uint Logical_Address = 0U;
                    if (sector_index > 0L) {
                        for (uint i = 0U, loopTo = (uint)(sector_index - 1L); i <= loopTo; i++) {
                            uint s_size = SectorSize(i);
                            Logical_Address += s_size;
                        }
                    }

                    bool Result = false;
                    try {
                        Result = FCUSB.USB_CONTROL_MSG_OUT(FlashcatUSB.USB.USBREQ.EXPIO_SECTORERASE, null, Logical_Address);
                    } catch (Exception ex) {
                    }

                    if (!Result)
                        return false;
                    FCUSB.USB_CONTROL_MSG_OUT(FlashcatUSB.USB.USBREQ.EXPIO_WAIT); // Calls the assigned WAIT function (uS, mS, SR, DQ7)
                    FCUSB.USB_WaitForComplete(); // Checks for WAIT flag to clear
                    bool blank_result = false;
                    uint timeout = 0U;
                    while (!blank_result) {
                        byte sr = GetStatusRegister();
                        if ((sr >> 7 & 1) == 1)
                            break;
                        timeout = (uint)(timeout + 1L);
                        if (timeout == 10L)
                            return false;
                        if (!blank_result)
                            FlashcatUSB.Utilities.Main.Sleep(100);
                    }
                }
            } catch {
            }

            return true;
        }

        public bool WriteData(long logical_address, byte[] data_to_write, FlashcatUSB.WriteParameters Params = null) {
            try {
                bool ReturnValue;
                uint DataToWrite = (uint)data_to_write.Length;
                uint PacketSize = 8192U; // Possibly /2 for IsFlashX8Mode
                int Loops = (int)Math.Ceiling(DataToWrite / (double)PacketSize); // Calcuates iterations
                for (int i = 0, loopTo = Loops - 1; i <= loopTo; i++) {
                    int BufferSize = (int)DataToWrite;
                    if (BufferSize > PacketSize)
                        BufferSize = (int)PacketSize;
                    var data = new byte[BufferSize];
                    Array.Copy(data_to_write, i * PacketSize, data, 0L, data.Length);
                    ReturnValue = WriteBulk_NOR((uint)logical_address, data);
                    if (!ReturnValue)
                        return false;
                    logical_address += data.Length;
                    DataToWrite = (uint)(DataToWrite - data.Length);
                    FCUSB.USB_WaitForComplete();
                }

                WaitUntilReady();
            } catch (Exception ex) {
            }

            return true;
        }

        public void WaitUntilReady() {
            try {
                FCUSB.USB_CONTROL_MSG_OUT(FlashcatUSB.USB.USBREQ.EXPIO_WAIT);
                FCUSB.USB_WaitForComplete(); // Checks for WAIT flag to clear
            } catch (Exception ex) {
            }
        }

        public void WaitForReady() => WaitUntilReady();

        public byte GetStatusRegister() {
            var d = new byte[1];
            if (!FCUSB.USB_CONTROL_MSG_IN(FlashcatUSB.USB.USBREQ.EXPIO_SR, ref d))
                return 0;
            return d[0];
        }

        public long SectorFind(uint sector_index, FlashcatUSB.FlashMemory.FlashArea memory_area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            uint base_addr = 0U;
            if (sector_index > 0L) {
                for (uint i = 0U, loopTo = (uint)(sector_index - 1L); i <= loopTo; i++)
                    base_addr += SectorSize(i, memory_area);
            }

            return base_addr;
        }

        public bool SectorWrite(uint sector_index, byte[] data, FlashcatUSB.WriteParameters Params = null) {
            uint Addr32 = (uint)this.SectorFind(sector_index, Params.Memory_Area);
            return WriteData(Addr32, data, Params);
        }

        public uint SectorCount() {
            return MyFlashDevice.Sector_Count;
        }

        public bool EraseDevice() {
            try {
                try {
                    FCUSB.USB_CONTROL_MSG_OUT(FlashcatUSB.USB.USBREQ.EXPIO_CHIPERASE);
                    FlashcatUSB.Utilities.Main.Sleep(1000); // Perform blank check
                    for (int i = 0; i <= 119; i++) { // 2 minutes
                        byte sr = GetStatusRegister();
                        if ((sr >> 7 & 1) == 1)
                            return true;
                        FlashcatUSB.Utilities.Main.Sleep(1000);
                    }

                    return false; // Timeout (device erase failed)
                } catch (Exception ex) {
                }
            } catch (Exception ex) {
            }

            return false;
        }

        internal uint SectorSize(uint sector, FlashcatUSB.FlashMemory.FlashArea memory_area = FlashcatUSB.FlashMemory.FlashArea.Main) {
            return MyFlashDevice.SECTOR_SIZE;
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        /* TODO ERROR: Skipped RegionDirectiveTrivia */
        private bool EXPIO_SETUP_USB(FlashcatUSB.MEM_PROTOCOL mode) {
            try {
                var result_data = new byte[1];
                bool result = FCUSB.USB_CONTROL_MSG_IN(FlashcatUSB.USB.USBREQ.EXPIO_INIT, ref result_data, (uint)((int)mode | 10 << 16));
                return result;
            } catch (Exception ex) {
            }

            return false;
        }

        /* TODO ERROR: Skipped EndRegionDirectiveTrivia */
        private FlashcatUSB.FlashMemory.FlashDetectResult DetectFlash() {
            var result = new FlashcatUSB.FlashMemory.FlashDetectResult();
            result.Successful = false;
            try {
                var ident_data = new byte[8]; // Contains 8 bytes
                if (!FCUSB.USB_CONTROL_MSG_IN(FlashcatUSB.USB.USBREQ.EXPIO_RDID, ref ident_data))
                    return result;
                if (ident_data[0] == 0 && ident_data[2] == 0)
                    return result; // 0x0000
                if (ident_data[0] == 0x90 && ident_data[2] == 0x90)
                    return result; // 0x9090 
                if (ident_data[0] == 0x90 && ident_data[2] == 0)
                    return result; // 0x9000 
                if (ident_data[0] == 0xFF && ident_data[2] == 0xFF)
                    return result; // 0xFFFF 
                if (ident_data[0] == 0xFF && ident_data[2] == 0)
                    return result; // 0xFF00
                if (ident_data[0] == 0x1 && ident_data[1] == 0 && ident_data[2] == 0x1 && ident_data[3] == 0)
                    return result; // 0x01000100
                result.MFG = ident_data[0];
                result.ID1 = (ushort)((uint)ident_data[1] << 8 | ident_data[2]);
                result.ID2 = (ushort)((uint)ident_data[3] << 8 | ident_data[4]);
                if ((int)result.ID1 == 0 && (int)result.ID2 == 0)
                    return result;
                result.Successful = true;
            } catch (Exception ex) {
            }

            return result;
        }

        private byte[] GetSetupPacket(uint Address, uint Count, ushort PageSize) {
            byte addr_bytes = 0;
            var data_in = new byte[20]; // 18 bytes total
            data_in[0] = (byte)(Address & 255L);
            data_in[1] = (byte)(Address >> 8 & 255L);
            data_in[2] = (byte)(Address >> 16 & 255L);
            data_in[3] = (byte)(Address >> 24 & 255L);
            data_in[4] = (byte)(Count & 255L);
            data_in[5] = (byte)(Count >> 8 & 255L);
            data_in[6] = (byte)(Count >> 16 & 255L);
            data_in[7] = (byte)(Count >> 24 & 255L);
            data_in[8] = (byte)(PageSize & 255); // This is how many bytes to increment between operations
            data_in[9] = (byte)(PageSize >> 8 & 255);
            return data_in;
        }

        private byte[] ReadBulk_NOR(uint address, uint count) {
            try {
                uint read_count = count;
                bool addr_offset = false;
                bool count_offset = false;
                if (address % 2L == 1L) {
                    addr_offset = true;
                    address = (uint)(address - 1L);
                    read_count = (uint)(read_count + 1L);
                }

                if (read_count % 2L == 1L) {
                    count_offset = true;
                    read_count = (uint)(read_count + 1L);
                }

                var setup_data = GetSetupPacket(address, read_count, (ushort)MyFlashDevice.PAGE_SIZE);
                var data_out = new byte[(int)(read_count - 1L + 1)]; // Bytes we want to read
                bool result = FCUSB.USB_SETUP_BULKIN(FlashcatUSB.USB.USBREQ.EXPIO_READDATA, setup_data, ref data_out, 0U);
                if (!result)
                    return null;
                if (addr_offset) {
                    var new_data = new byte[(int)(count - 1L + 1)];
                    Array.Copy(data_out, 1, new_data, 0, new_data.Length);
                    data_out = new_data;
                } else {
                    Array.Resize(ref data_out, (int)(count - 1L + 1));
                }

                return data_out;
            } catch (Exception ex) {
            }

            return null;
        }

        private bool WriteBulk_NOR(uint address, byte[] data_out) {
            try {
                var setup_data = GetSetupPacket(address, (uint)data_out.Length, (ushort)MyFlashDevice.PAGE_SIZE);
                bool result = FCUSB.USB_SETUP_BULKOUT(FlashcatUSB.USB.USBREQ.EXPIO_WRITEDATA, setup_data, data_out, 0U);
                return result;
            } catch (Exception ex) {
            }

            return false;
        }
    }
}

File: Windows/programmers/HyperFlash.vb
Imports FlashcatUSB.PARALLEL_NOR
Imports FlashcatUSB.FlashMemory
Imports FlashcatUSB.USB
Imports FlashcatUSB.USB.HostClient

Public Class HF_Programmer : Implements MemoryDeviceUSB
    Private FCUSB As FCUSB_DEVICE
    Public Property MyFlashDevice As HYPERFLASH  'Contains the definition for the parallel mode device that is connected
    Public Property MyFlashStatus As DeviceStatus = DeviceStatus.NotDetected

    Public Event PrintConsole(msg As String) Implements MemoryDeviceUSB.PrintConsole
    Public Event SetProgress(percent As Integer) Implements MemoryDeviceUSB.SetProgress 'Not used

    Sub New(parent_if As FCUSB_DEVICE)
        FCUSB = parent_if
    End Sub

    Public Function DeviceInit() As Boolean Implements MemoryDeviceUSB.DeviceInit
        FCUSB.USB_VCC_OFF()
        MyFlashDevice = Nothing
        EXPIO_SETUP_USB(MEM_PROTOCOL.HYPERFLASH)
        FCUSB.USB_VCC_ON(MySettings.VOLT_SELECT)
        Utilities.Sleep(200)
        Dim HF_DETECT As FlashDetectResult = DetectFlash()
        If HF_DETECT.Successful Then
            Dim chip_id_str As String = Hex(HF_DETECT.MFG).PadLeft(2, "0"c) & Hex(HF_DETECT.ID1).PadLeft(8, "0"c)
            RaiseEvent PrintConsole(String.Format(RM.GetString("ext_connected_chipid"), chip_id_str))
            Dim device_matches() As Device = FlashDatabase.FindDevices(HF_DETECT.MFG, HF_DETECT.ID1, 0, MemoryType.HYPERFLASH)
            If (device_matches IsNot Nothing AndAlso device_matches.Length > 0) Then
                If MyFlashDevice Is Nothing Then MyFlashDevice = CType(device_matches(0), HYPERFLASH)
                RaiseEvent PrintConsole(String.Format(RM.GetString("flash_detected"), MyFlashDevice.NAME, Format(MyFlashDevice.FLASH_SIZE, "#,###")))
                MyFlashStatus = DeviceStatus.Supported
            Else
                MyFlashStatus = DeviceStatus.NotSupported
            End If
            Return True
        Else
            MyFlashStatus = DeviceStatus.NotDetected
            Return False
        End If
    End Function

#Region "Public Interface"

    Public ReadOnly Property GetDevice As Device Implements MemoryDeviceUSB.GetDevice
        Get
            Return Me.MyFlashDevice
        End Get
    End Property

    Friend ReadOnly Property DeviceName() As String Implements MemoryDeviceUSB.DeviceName
        Get
            Select Case MyFlashStatus
                Case DeviceStatus.Supported
                    Return MyFlashDevice.NAME
                Case DeviceStatus.NotSupported
                    Return Hex(MyFlashDevice.MFG_CODE).PadLeft(2, "0"c) & " " & Hex(MyFlashDevice.ID1).PadLeft(4, "0"c)
                Case Else
                    Return RM.GetString("no_flash_detected")
            End Select
        End Get
    End Property

    Friend ReadOnly Property DeviceSize As Long Implements MemoryDeviceUSB.DeviceSize
        Get
            Return MyFlashDevice.FLASH_SIZE
        End Get
    End Property

    Public Function ReadData(logical_address As Long, data_count As Integer) As Byte() Implements MemoryDeviceUSB.ReadData
        Return ReadBulk_NOR(CUInt(logical_address), CInt(data_count))
    End Function

    Public Function SectorErase(sector_index As Integer) As Boolean Implements MemoryDeviceUSB.SectorErase
        Try
            If sector_index = 0 AndAlso SectorSize(0) = MyFlashDevice.FLASH_SIZE Then
                Return EraseDevice() 'Single sector, must do a full chip erase instead
            Else
                Dim Logical_Address As UInt32 = 0
                If (sector_index > 0) Then
                    For i As Integer = 0 To sector_index - 1
                        Dim s_size As Integer = SectorSize(i)
                        Logical_Address += CUInt(s_size)
                    Next
                End If
                Dim Result As Boolean = False
                Try
                    Result = FCUSB.USB_CONTROL_MSG_OUT(USBREQ.EXPIO_SECTORERASE, Nothing, Logical_Address)
                Catch ex As Exception
                End Try
                If Not Result Then Return False
                FCUSB.USB_CONTROL_MSG_OUT(USBREQ.EXPIO_WAIT) 'Calls the assigned WAIT function (uS, mS, SR, DQ7)
                FCUSB.USB_WaitForComplete() 'Checks for WAIT flag to clear
                Dim blank_result As Boolean = False
                Dim timeout As UInt32 = 0
                Do Until blank_result
                    Dim sr As Byte = GetStatusRegister()
                    If ((sr >> 7) And 1) = 1 Then Exit Do
                    timeout += 1UI
                    If (timeout = 10) Then Return False
                    If Not blank_result Then Utilities.Sleep(100)
                Loop
            End If
        Catch
        End Try
        Return True
    End Function

    Public Function WriteData(logical_address As Long, data_to_write() As Byte, Optional Params As WriteParameters = Nothing) As Boolean Implements MemoryDeviceUSB.WriteData
        Try
            Dim ReturnValue As Boolean
            Dim DataToWrite As Integer = data_to_write.Length
            Dim PacketSize As Integer = 8192 'Possibly /2 for IsFlashX8Mode
            Dim Loops As Integer = CInt(Math.Ceiling(DataToWrite / PacketSize)) 'Calcuates iterations
            For i As Integer = 0 To Loops - 1
                Dim BufferSize As Integer = DataToWrite
                If (BufferSize > PacketSize) Then BufferSize = PacketSize
                Dim data(BufferSize - 1) As Byte
                Array.Copy(data_to_write, (i * PacketSize), data, 0, data.Length)
                ReturnValue = WriteBulk_NOR(CUInt(logical_address), data)
                If (Not ReturnValue) Then Return False
                logical_address += data.Length
                DataToWrite -= data.Length
                FCUSB.USB_WaitForComplete()
            Next
            WaitUntilReady()
        Catch ex As Exception
        End Try
        Return True
    End Function

    Public Sub WaitUntilReady() Implements MemoryDeviceUSB.WaitUntilReady
        Try
            FCUSB.USB_CONTROL_MSG_OUT(USBREQ.EXPIO_WAIT)
            FCUSB.USB_WaitForComplete() 'Checks for WAIT flag to clear
        Catch ex As Exception
        End Try
    End Sub

    Public Function GetStatusRegister() As Byte
        Dim d(0) As Byte
        If Not FCUSB.USB_CONTROL_MSG_IN(USBREQ.EXPIO_SR, d) Then Return 0
        Return d(0)
    End Function

    Public Function SectorFind(sector_index As Integer) As Long Implements MemoryDeviceUSB.SectorFind
        Dim base_addr As UInt32 = 0
        If sector_index > 0 Then
            For i As Integer = 0 To sector_index - 1
                base_addr += CUInt(Me.SectorSize(i))
            Next
        End If
        Return base_addr
    End Function

    Public Function SectorWrite(sector_index As Integer, data() As Byte, Optional Params As WriteParameters = Nothing) As Boolean Implements MemoryDeviceUSB.SectorWrite
        Dim Addr32 As Long = Me.SectorFind(sector_index)
        Return WriteData(Addr32, data, Params)
    End Function

    Public Function SectorCount() As Integer Implements MemoryDeviceUSB.SectorCount
        Return MyFlashDevice.Sector_Count
    End Function

    Public Function EraseDevice() As Boolean Implements MemoryDeviceUSB.EraseDevice
        Try
            FCUSB.USB_CONTROL_MSG_OUT(USBREQ.EXPIO_CHIPERASE)
            Utilities.Sleep(1000) 'Perform blank check
            For i = 0 To 119 '2 minutes
                Dim sr As Byte = GetStatusRegister()
                If ((sr >> 7) And 1) = 1 Then Return True
                Utilities.Sleep(1000)
            Next
            Return False 'Timeout (device erase failed)
        Catch ex As Exception
        End Try
        Return False
    End Function

    Friend Function SectorSize(sector As Integer) As Integer Implements MemoryDeviceUSB.SectorSize
        Return MyFlashDevice.SECTOR_SIZE
    End Function

#End Region

#Region "EXPIO_SETUP"

    Private Function EXPIO_SETUP_USB(mode As MEM_PROTOCOL) As Boolean
        Try
            Dim result_data(0) As Byte
            Dim result As Boolean = FCUSB.USB_CONTROL_MSG_IN(USBREQ.EXPIO_INIT, result_data, CUInt(mode Or (10 << 16)))
            Return result
        Catch ex As Exception
        End Try
        Return False
    End Function

#End Region

    Private Function DetectFlash() As FlashDetectResult
        Dim result As New FlashDetectResult
        result.Successful = False
        Try
            Dim ident_data(7) As Byte 'Contains 8 bytes
            If Not FCUSB.USB_CONTROL_MSG_IN(USBREQ.EXPIO_RDID, ident_data) Then Return result
            If ident_data(0) = 0 AndAlso ident_data(2) = 0 Then Return result '0x0000
            If ident_data(0) = &H90 AndAlso ident_data(2) = &H90 Then Return result '0x9090 
            If ident_data(0) = &H90 AndAlso ident_data(2) = 0 Then Return result '0x9000 
            If ident_data(0) = &HFF AndAlso ident_data(2) = &HFF Then Return result '0xFFFF 
            If ident_data(0) = &HFF AndAlso ident_data(2) = 0 Then Return result '0xFF00
            If ident_data(0) = &H1 AndAlso ident_data(1) = 0 AndAlso ident_data(2) = &H1 AndAlso ident_data(3) = 0 Then Return result '0x01000100
            result.MFG = ident_data(0)
            result.ID1 = (CUShort(ident_data(1)) << 8) Or CUShort(ident_data(2))
            result.ID2 = (CUShort(ident_data(3)) << 8) Or CUShort(ident_data(4))
            If result.ID1 = 0 AndAlso result.ID2 = 0 Then Return result
            result.Successful = True
        Catch ex As Exception
        End Try
        Return result
    End Function

    Private Function GetSetupPacket(Address As UInt32, Count As UInt32, PageSize As UInt16) As Byte()
        Dim data_in(19) As Byte '18 bytes total
        data_in(0) = CByte(Address And 255)
        data_in(1) = CByte((Address >> 8) And 255)
        data_in(2) = CByte((Address >> 16) And 255)
        data_in(3) = CByte((Address >> 24) And 255)
        data_in(4) = CByte(Count And 255)
        data_in(5) = CByte((Count >> 8) And 255)
        data_in(6) = CByte((Count >> 16) And 255)
        data_in(7) = CByte((Count >> 24) And 255)
        data_in(8) = CByte(PageSize And 255) 'This is how many bytes to increment between operations
        data_in(9) = CByte((PageSize >> 8) And 255)
        Return data_in
    End Function

    Private Function ReadBulk_NOR(address As UInt32, count As Integer) As Byte()
        Try
            Dim read_count As Integer = count
            Dim addr_offset As Boolean = False
            If (address Mod 2 = 1) Then
                addr_offset = True
                address = (address - 1UI)
                read_count += 1
            End If
            If (read_count Mod 2 = 1) Then
                read_count += 1
            End If
            Dim setup_data() As Byte = GetSetupPacket(address, CUInt(read_count), MyFlashDevice.PAGE_SIZE)
            Dim data_out(read_count - 1) As Byte 'Bytes we want to read
            Dim result As Boolean = FCUSB.USB_SETUP_BULKIN(USBREQ.EXPIO_READDATA, setup_data, data_out, 0)
            If Not result Then Return Nothing
            If addr_offset Then
                Dim new_data(count - 1) As Byte
                Array.Copy(data_out, 1, new_data, 0, new_data.Length)
                data_out = new_data
            Else
                ReDim Preserve data_out(count - 1)
            End If
            Return data_out
        Catch ex As Exception
        End Try
        Return Nothing
    End Function

    Private Function WriteBulk_NOR(address As UInt32, data_out() As Byte) As Boolean
        Try
            Dim setup_data() As Byte = GetSetupPacket(address, CUInt(data_out.Length), MyFlashDevice.PAGE_SIZE)
            Dim result As Boolean = FCUSB.USB_SETUP_BULKOUT(USBREQ.EXPIO_WRITEDATA, setup_data, data_out, 0)
            Return result
        Catch ex As Exception
        End Try
        Return False
    End Function

End Class
